File: agenkit/safety/audit.py
"""
Security audit logging.

Provides comprehensive security event logging:
- Access attempts
- Permission checks
- Validation failures
- Anomaly detections
- Security violations
"""

import json
import logging
import logging.handlers
from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import Enum
from typing import Any


class AuditEventType(Enum):
    """Types of audit events."""

    # Access events
    ACCESS_GRANTED = "access_granted"
    ACCESS_DENIED = "access_denied"

    # Validation events
    INPUT_VALIDATION_FAILED = "input_validation_failed"
    OUTPUT_VALIDATION_FAILED = "output_validation_failed"

    # Permission events
    PERMISSION_GRANTED = "permission_granted"
    PERMISSION_DENIED = "permission_denied"

    # Security events
    PROMPT_INJECTION_DETECTED = "prompt_injection_detected"
    SENSITIVE_DATA_DETECTED = "sensitive_data_detected"
    ANOMALY_DETECTED = "anomaly_detected"

    # Operational events
    AGENT_STARTED = "agent_started"
    AGENT_COMPLETED = "agent_completed"
    AGENT_FAILED = "agent_failed"


class AuditSeverity(Enum):
    """Severity levels for audit events."""

    INFO = "info"
    WARNING = "warning"
    ERROR = "error"
    CRITICAL = "critical"


@dataclass
class AuditEvent:
    """
    Structured audit event.

    Contains all information needed for security auditing and compliance.
    """

    event_type: AuditEventType
    severity: AuditSeverity
    timestamp: str = field(default_factory=lambda: datetime.now(timezone.utc).isoformat())
    user_id: str | None = None
    agent_name: str | None = None
    message: str | None = None
    details: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        """Convert to dictionary for serialization."""
        return {
            "event_type": self.event_type.value,
            "severity": self.severity.value,
            "timestamp": self.timestamp,
            "user_id": self.user_id,
            "agent_name": self.agent_name,
            "message": self.message,
            "details": self.details,
        }

    def to_json(self) -> str:
        """Convert to JSON string."""
        return json.dumps(self.to_dict())


class SecurityAuditLogger:
    """
    Security audit logger with structured logging.

    Features:
    - Structured JSON logging
    - Log rotation
    - Severity-based filtering
    - Multiple output targets
    - Searchable audit trail
    """

    def __init__(
        self,
        log_file: str = "security_audit.log",
        max_bytes: int = 100 * 1024 * 1024,  # 100MB
        backup_count: int = 10,
        min_severity: AuditSeverity = AuditSeverity.INFO,
        also_log_to_console: bool = True,
    ):
        """
        Initialize security audit logger.

        Args:
            log_file: Path to log file
            max_bytes: Maximum log file size before rotation
            backup_count: Number of backup files to keep
            min_severity: Minimum severity to log
            also_log_to_console: Also output to console
        """
        self.min_severity = min_severity
        self.logger = logging.getLogger("agenkit.security.audit")
        self.logger.setLevel(logging.INFO)

        # Clear existing handlers
        self.logger.handlers = []

        # File handler with rotation
        file_handler = logging.handlers.RotatingFileHandler(
            log_file, maxBytes=max_bytes, backupCount=backup_count
        )

        # JSON formatter
        formatter = logging.Formatter("%(message)s")
        file_handler.setFormatter(formatter)
        self.logger.addHandler(file_handler)

        # Console handler (optional)
        if also_log_to_console:
            console_handler = logging.StreamHandler()
            console_handler.setFormatter(formatter)
            self.logger.addHandler(console_handler)

    def _should_log(self, severity: AuditSeverity) -> bool:
        """Check if event should be logged based on severity."""
        severity_order = {
            AuditSeverity.INFO: 0,
            AuditSeverity.WARNING: 1,
            AuditSeverity.ERROR: 2,
            AuditSeverity.CRITICAL: 3,
        }
        return severity_order[severity] >= severity_order[self.min_severity]

    def log(self, event: AuditEvent):
        """
        Log audit event.

        Args:
            event: Audit event to log
        """
        if not self._should_log(event.severity):
            return

        # Map severity to logging level
        level_map = {
            AuditSeverity.INFO: logging.INFO,
            AuditSeverity.WARNING: logging.WARNING,
            AuditSeverity.ERROR: logging.ERROR,
            AuditSeverity.CRITICAL: logging.CRITICAL,
        }

        # Log as JSON
        self.logger.log(level_map[event.severity], event.to_json())

    def log_access(
        self,
        granted: bool,
        user_id: str,
        agent_name: str,
        action: str,
        details: dict[str, Any] | None = None,
    ):
        """Log access attempt."""
        event = AuditEvent(
            event_type=AuditEventType.ACCESS_GRANTED if granted else AuditEventType.ACCESS_DENIED,
            severity=AuditSeverity.INFO if granted else AuditSeverity.WARNING,
            user_id=user_id,
            agent_name=agent_name,
            message=f"Access {('granted' if granted else 'denied')} for action: {action}",
            details=details or {},
        )
        self.log(event)

    def log_permission_check(
        self,
        granted: bool,
        user_id: str,
        agent_name: str,
        permission: str,
        details: dict[str, Any] | None = None,
    ):
        """Log permission check."""
        event = AuditEvent(
            event_type=AuditEventType.PERMISSION_GRANTED
            if granted
            else AuditEventType.PERMISSION_DENIED,
            severity=AuditSeverity.INFO if granted else AuditSeverity.WARNING,
            user_id=user_id,
            agent_name=agent_name,
            message=f"Permission {permission}: {('granted' if granted else 'denied')}",
            details=details or {},
        )
        self.log(event)

    def log_validation_failure(
        self,
        user_id: str,
        validation_type: str,
        reason: str,
        content_preview: str | None = None,
        agent_name: str | None = None,
    ):
        """Log validation failure."""
        # Truncate content preview
        truncated_preview = ""
        if content_preview:
            truncated_preview = content_preview[:200] + (
                "..." if len(content_preview) > 200 else ""
            )

        event = AuditEvent(
            event_type=AuditEventType.INPUT_VALIDATION_FAILED
            if validation_type == "input"
            else AuditEventType.OUTPUT_VALIDATION_FAILED,
            severity=AuditSeverity.ERROR,
            user_id=user_id,
            agent_name=agent_name or "",
            message=f"{validation_type.capitalize()} validation failed: {reason}",
            details={
                "validation_type": validation_type,
                "reason": reason,
                "content_preview": truncated_preview,
            },
        )
        self.log(event)

    def log_prompt_injection(
        self,
        user_id: str,
        score: int,
        matched_patterns: list,
        content_preview: str | None = None,
        agent_name: str | None = None,
    ):
        """Log prompt injection detection."""
        # Truncate content preview
        truncated_preview = ""
        if content_preview:
            truncated_preview = content_preview[:200] + (
                "..." if len(content_preview) > 200 else ""
            )

        event = AuditEvent(
            event_type=AuditEventType.PROMPT_INJECTION_DETECTED,
            severity=AuditSeverity.ERROR,
            user_id=user_id,
            agent_name=agent_name or "",
            message=f"Prompt injection detected (score: {score}, patterns: {len(matched_patterns)})",
            details={
                "score": score,
                "matched_patterns": matched_patterns,
                "content_preview": truncated_preview,
            },
        )
        self.log(event)

    def log_anomaly(
        self,
        user_id: str,
        anomaly_type: str,
        details: dict[str, Any] | None = None,
        agent_name: str | None = None,
    ):
        """Log anomaly detection."""
        event_details = details or {}
        event_details["anomaly_type"] = anomaly_type

        event = AuditEvent(
            event_type=AuditEventType.ANOMALY_DETECTED,
            severity=AuditSeverity.WARNING,
            user_id=user_id,
            agent_name=agent_name or "",
            message=f"Anomaly detected: {anomaly_type}",
            details=event_details,
        )
        self.log(event)

    def log_agent_execution(
        self,
        user_id: str,
        agent_name: str,
        status: str,  # "started", "completed", "failed"
        duration: float | None = None,
        error: str | None = None,
        details: dict[str, Any] | None = None,
    ):
        """Log agent execution."""
        event_type_map = {
            "started": AuditEventType.AGENT_STARTED,
            "completed": AuditEventType.AGENT_COMPLETED,
            "failed": AuditEventType.AGENT_FAILED,
        }

        severity_map = {
            "started": AuditSeverity.INFO,
            "completed": AuditSeverity.INFO,
            "failed": AuditSeverity.ERROR,
        }

        event_details = details or {}
        if duration is not None:
            event_details["duration_seconds"] = duration
        if error:
            event_details["error"] = error

        event = AuditEvent(
            event_type=event_type_map[status],
            severity=severity_map[status],
            user_id=user_id,
            agent_name=agent_name,
            message=f"Agent {status}" + (f" ({duration:.2f}s)" if duration else ""),
            details=event_details,
        )
        self.log(event)

    # Convenience methods for common logging scenarios

    def log_access_granted(
        self,
        user_id: str,
        resource: str,
        permission: str,
        agent_name: str | None = None,
    ):
        """Log successful access grant."""
        event = AuditEvent(
            event_type=AuditEventType.ACCESS_GRANTED,
            severity=AuditSeverity.INFO,
            user_id=user_id,
            agent_name=agent_name or "",
            message=f"Access granted to resource: {resource}",
            details={"resource": resource, "permission": permission},
        )
        self.log(event)

    def log_access_denied(
        self,
        user_id: str,
        resource: str,
        permission: str,
        reason: str,
        agent_name: str | None = None,
    ):
        """Log access denial."""
        event = AuditEvent(
            event_type=AuditEventType.ACCESS_DENIED,
            severity=AuditSeverity.WARNING,
            user_id=user_id,
            agent_name=agent_name or "",
            message=f"Access denied to resource: {resource}",
            details={"resource": resource, "permission": permission, "reason": reason},
        )
        self.log(event)

    def log_sensitive_data_redaction(
        self,
        user_id: str,
        fields_redacted: list,
        output_preview: str,
        agent_name: str | None = None,
    ):
        """Log sensitive data redaction."""
        # Truncate output preview
        truncated_preview = output_preview[:200] + ("..." if len(output_preview) > 200 else "")

        event = AuditEvent(
            event_type=AuditEventType.SENSITIVE_DATA_DETECTED,
            severity=AuditSeverity.WARNING,
            user_id=user_id,
            agent_name=agent_name or "",
            message=f"Sensitive data redacted: {len(fields_redacted)} field(s)",
            details={
                "fields_redacted": fields_redacted,
                "output_preview": truncated_preview,
            },
        )
        self.log(event)


# Global audit logger instance (can be configured once)
_global_audit_logger: SecurityAuditLogger | None = None


def get_audit_logger() -> SecurityAuditLogger:
    """
    Get global audit logger instance.

    Returns:
        SecurityAuditLogger instance
    """
    global _global_audit_logger
    if _global_audit_logger is None:
        _global_audit_logger = SecurityAuditLogger()
    return _global_audit_logger


def configure_audit_logger(
    log_file: str = "security_audit.log",
    max_bytes: int = 100 * 1024 * 1024,
    backup_count: int = 10,
    min_severity: AuditSeverity = AuditSeverity.INFO,
    also_log_to_console: bool = True,
):
    """
    Configure global audit logger.

    Args:
        log_file: Path to log file
        max_bytes: Maximum log file size before rotation
        backup_count: Number of backup files to keep
        min_severity: Minimum severity to log
        also_log_to_console: Also output to console
    """
    global _global_audit_logger
    _global_audit_logger = SecurityAuditLogger(
        log_file, max_bytes, backup_count, min_severity, also_log_to_console
    )

File: agenkit-ts/src/safety/audit.ts
/**
 * Security audit logging for agent operations.
 *
 * Provides structured logging with:
 * - Event type classification
 * - Severity levels
 * - JSON formatting
 * - Log rotation support
 * - Searchable audit trail
 */

import * as fs from 'fs';
import * as path from 'path';

/**
 * Types of security audit events.
 */
export enum AuditEventType {
  ACCESS_GRANTED = 'access_granted',
  ACCESS_DENIED = 'access_denied',
  PERMISSION_GRANTED = 'permission_granted',
  PERMISSION_DENIED = 'permission_denied',
  INPUT_VALIDATION_FAILED = 'input_validation_failed',
  OUTPUT_VALIDATION_FAILED = 'output_validation_failed',
  PROMPT_INJECTION_DETECTED = 'prompt_injection_detected',
  SENSITIVE_DATA_DETECTED = 'sensitive_data_detected',
  ANOMALY_DETECTED = 'anomaly_detected',
  AGENT_STARTED = 'agent_started',
  AGENT_COMPLETED = 'agent_completed',
  AGENT_FAILED = 'agent_failed',
}

/**
 * Severity levels for audit events.
 */
export enum AuditSeverity {
  INFO = 'info',
  WARNING = 'warning',
  ERROR = 'error',
  CRITICAL = 'critical',
}

/**
 * Structured audit event.
 */
export interface AuditEvent {
  /** Type of audit event */
  eventType: AuditEventType;

  /** Event severity */
  severity: AuditSeverity;

  /** User identifier */
  userId: string;

  /** Agent name (if applicable) */
  agentName: string;

  /** Human-readable message */
  message: string;

  /** ISO 8601 timestamp */
  timestamp: string;

  /** Additional event details */
  details: Record<string, unknown>;
}

/**
 * Configuration for audit logger.
 */
export interface SecurityAuditLoggerConfig {
  /** Path to log file (default: 'security_audit.log') */
  logFile?: string;

  /** Maximum log file size before rotation in bytes (default: 100MB) */
  maxBytes?: number;

  /** Number of backup files to keep (default: 10) */
  backupCount?: number;

  /** Minimum severity to log (default: INFO) */
  minSeverity?: AuditSeverity;

  /** Also output to console (default: true) */
  alsoLogToConsole?: boolean;
}

/**
 * Security audit logger with structured logging and log rotation.
 *
 * Features:
 * - Structured JSON logging
 * - Log rotation based on file size
 * - Severity-based filtering
 * - Console and file output
 * - Searchable audit trail
 *
 * Example:
 *   const logger = new SecurityAuditLogger({
 *     logFile: 'audit.log',
 *     minSeverity: AuditSeverity.WARNING,
 *   });
 *
 *   logger.logAccessDenied('user_123', 'admin_panel', 'read', 'Insufficient permissions');
 */
export class SecurityAuditLogger {
  private logFile: string;
  private maxBytes: number;
  private backupCount: number;
  private minSeverity: AuditSeverity;
  private alsoLogToConsole: boolean;
  private writeStream: fs.WriteStream | null = null;
  private currentSize: number = 0;

  constructor(config?: SecurityAuditLoggerConfig) {
    this.logFile = config?.logFile ?? 'security_audit.log';
    this.maxBytes = config?.maxBytes ?? 100 * 1024 * 1024; // 100MB
    this.backupCount = config?.backupCount ?? 10;
    this.minSeverity = config?.minSeverity ?? AuditSeverity.INFO;
    this.alsoLogToConsole = config?.alsoLogToConsole ?? true;

    // Create log file if it doesn't exist and get current size
    if (fs.existsSync(this.logFile)) {
      const stats = fs.statSync(this.logFile);
      this.currentSize = stats.size;
    }

    // Open write stream
    this.writeStream = fs.createWriteStream(this.logFile, { flags: 'a' });
  }

  /**
   * Check if severity should be logged based on minimum severity.
   */
  private shouldLog(severity: AuditSeverity): boolean {
    const severityOrder: Record<AuditSeverity, number> = {
      [AuditSeverity.INFO]: 0,
      [AuditSeverity.WARNING]: 1,
      [AuditSeverity.ERROR]: 2,
      [AuditSeverity.CRITICAL]: 3,
    };

    return severityOrder[severity] >= severityOrder[this.minSeverity];
  }

  /**
   * Rotate log files when size limit reached.
   */
  private rotateLog(): void {
    // Close current stream
    if (this.writeStream) {
      this.writeStream.end();
    }

    // Rotate backup files
    for (let i = this.backupCount - 1; i >= 1; i--) {
      const oldPath = `${this.logFile}.${i}`;
      const newPath = `${this.logFile}.${i + 1}`;
      if (fs.existsSync(oldPath)) {
        fs.renameSync(oldPath, newPath);
      }
    }

    // Move current log to .1
    if (fs.existsSync(this.logFile)) {
      fs.renameSync(this.logFile, `${this.logFile}.1`);
    }

    // Open new stream
    this.writeStream = fs.createWriteStream(this.logFile, { flags: 'a' });
    this.currentSize = 0;
  }

  /**
   * Create audit event object.
   */
  private createEvent(
    eventType: AuditEventType,
    severity: AuditSeverity,
    userId: string,
    agentName: string,
    message: string,
    details: Record<string, unknown>,
  ): AuditEvent {
    return {
      eventType,
      severity,
      userId,
      agentName,
      message,
      timestamp: new Date().toISOString(),
      details,
    };
  }

  /**
   * Log audit event.
   */
  log(event: AuditEvent): void {
    if (!this.shouldLog(event.severity)) {
      return;
    }

    const logLine = JSON.stringify(event) + '\n';
    const logBytes = Buffer.byteLength(logLine, 'utf8');

    // Check if rotation needed
    if (this.currentSize + logBytes > this.maxBytes) {
      this.rotateLog();
    }

    // Write to file
    if (this.writeStream) {
      this.writeStream.write(logLine);
      this.currentSize += logBytes;
    }

    // Also log to console if configured
    if (this.alsoLogToConsole) {
      const level = event.severity.toLowerCase();
      console.log(`[${level.toUpperCase()}] ${event.message}`, event.details);
    }
  }

  /**
   * Log access attempt.
   */
  logAccess(
    granted: boolean,
    userId: string,
    agentName: string,
    action: string,
    details?: Record<string, unknown>,
  ): void {
    const event = this.createEvent(
      granted ? AuditEventType.ACCESS_GRANTED : AuditEventType.ACCESS_DENIED,
      granted ? AuditSeverity.INFO : AuditSeverity.WARNING,
      userId,
      agentName,
      `Access ${granted ? 'granted' : 'denied'} for action: ${action}`,
      details || {},
    );
    this.log(event);
  }

  /**
   * Log permission check.
   */
  logPermissionCheck(
    granted: boolean,
    userId: string,
    agentName: string,
    permission: string,
    details?: Record<string, unknown>,
  ): void {
    const event = this.createEvent(
      granted ? AuditEventType.PERMISSION_GRANTED : AuditEventType.PERMISSION_DENIED,
      granted ? AuditSeverity.INFO : AuditSeverity.WARNING,
      userId,
      agentName,
      `Permission ${permission}: ${granted ? 'granted' : 'denied'}`,
      details || {},
    );
    this.log(event);
  }

  /**
   * Log validation failure.
   */
  logValidationFailure(
    userId: string,
    validationType: 'input' | 'output',
    reason: string,
    contentPreview?: string,
    agentName?: string,
  ): void {
    // Truncate content preview
    const truncatedPreview = contentPreview
      ? contentPreview.substring(0, 200) + (contentPreview.length > 200 ? '...' : '')
      : '';

    const event = this.createEvent(
      validationType === 'input'
        ? AuditEventType.INPUT_VALIDATION_FAILED
        : AuditEventType.OUTPUT_VALIDATION_FAILED,
      AuditSeverity.ERROR,
      userId,
      agentName || '',
      `${validationType.charAt(0).toUpperCase() + validationType.slice(1)} validation failed: ${reason}`,
      {
        validation_type: validationType,
        reason,
        content_preview: truncatedPreview,
      },
    );
    this.log(event);
  }

  /**
   * Log prompt injection detection.
   */
  logPromptInjection(
    userId: string,
    score: number,
    matchedPatterns: string[],
    contentPreview?: string,
    agentName?: string,
  ): void {
    // Truncate content preview
    const truncatedPreview = contentPreview
      ? contentPreview.substring(0, 200) + (contentPreview.length > 200 ? '...' : '')
      : '';

    const event = this.createEvent(
      AuditEventType.PROMPT_INJECTION_DETECTED,
      AuditSeverity.ERROR,
      userId,
      agentName || '',
      `Prompt injection detected (score: ${score}, patterns: ${matchedPatterns.length})`,
      {
        score,
        matched_patterns: matchedPatterns,
        content_preview: truncatedPreview,
      },
    );
    this.log(event);
  }

  /**
   * Log anomaly detection.
   */
  logAnomaly(
    userId: string,
    anomalyType: string,
    details?: Record<string, unknown>,
    agentName?: string,
  ): void {
    const eventDetails = { ...details, anomaly_type: anomalyType };

    const event = this.createEvent(
      AuditEventType.ANOMALY_DETECTED,
      AuditSeverity.WARNING,
      userId,
      agentName || '',
      `Anomaly detected: ${anomalyType}`,
      eventDetails,
    );
    this.log(event);
  }

  /**
   * Log agent execution.
   */
  logAgentExecution(
    userId: string,
    agentName: string,
    status: 'started' | 'completed' | 'failed',
    duration?: number,
    error?: string,
    details?: Record<string, unknown>,
  ): void {
    const eventTypeMap: Record<string, AuditEventType> = {
      started: AuditEventType.AGENT_STARTED,
      completed: AuditEventType.AGENT_COMPLETED,
      failed: AuditEventType.AGENT_FAILED,
    };

    const severityMap: Record<string, AuditSeverity> = {
      started: AuditSeverity.INFO,
      completed: AuditSeverity.INFO,
      failed: AuditSeverity.ERROR,
    };

    const eventDetails = { ...details };
    if (duration !== undefined) {
      eventDetails.duration_seconds = duration;
    }
    if (error) {
      eventDetails.error = error;
    }

    const message = `Agent ${status}${duration !== undefined ? ` (${duration.toFixed(2)}s)` : ''}`;

    const event = this.createEvent(
      eventTypeMap[status],
      severityMap[status],
      userId,
      agentName,
      message,
      eventDetails,
    );
    this.log(event);
  }

  /**
   * Log successful access grant.
   */
  logAccessGranted(
    userId: string,
    resource: string,
    permission: string,
    agentName?: string,
  ): void {
    const event = this.createEvent(
      AuditEventType.ACCESS_GRANTED,
      AuditSeverity.INFO,
      userId,
      agentName || '',
      `Access granted to resource: ${resource}`,
      { resource, permission },
    );
    this.log(event);
  }

  /**
   * Log access denial.
   */
  logAccessDenied(
    userId: string,
    resource: string,
    permission: string,
    reason: string,
    agentName?: string,
  ): void {
    const event = this.createEvent(
      AuditEventType.ACCESS_DENIED,
      AuditSeverity.WARNING,
      userId,
      agentName || '',
      `Access denied to resource: ${resource}`,
      { resource, permission, reason },
    );
    this.log(event);
  }

  /**
   * Log sensitive data redaction.
   */
  logSensitiveDataRedaction(
    userId: string,
    fieldsRedacted: string[],
    outputPreview: string,
    agentName?: string,
  ): void {
    // Truncate output preview
    const truncatedPreview =
      outputPreview.substring(0, 200) + (outputPreview.length > 200 ? '...' : '');

    const event = this.createEvent(
      AuditEventType.SENSITIVE_DATA_DETECTED,
      AuditSeverity.WARNING,
      userId,
      agentName || '',
      `Sensitive data redacted: ${fieldsRedacted.length} field(s)`,
      {
        fields_redacted: fieldsRedacted,
        output_preview: truncatedPreview,
      },
    );
    this.log(event);
  }

  /**
   * Close the logger and flush logs.
   * Returns a Promise that resolves when the stream is fully closed.
   */
  close(): Promise<void> {
    return new Promise((resolve) => {
      if (this.writeStream) {
        this.writeStream.end(() => {
          this.writeStream = null;
          resolve();
        });
      } else {
        resolve();
      }
    });
  }
}

// Global audit logger instance
let globalAuditLogger: SecurityAuditLogger | null = null;

/**
 * Get global audit logger instance.
 */
export function getAuditLogger(): SecurityAuditLogger {
  if (!globalAuditLogger) {
    globalAuditLogger = new SecurityAuditLogger();
  }
  return globalAuditLogger;
}

/**
 * Configure global audit logger.
 */
export async function configureAuditLogger(config: SecurityAuditLoggerConfig): Promise<void> {
  if (globalAuditLogger) {
    await globalAuditLogger.close();
  }
  globalAuditLogger = new SecurityAuditLogger(config);
}
